elyra/pipeline/kfp/kfp_authentication.py

```diff
@@ -0,0 +1,605 @@
+#
+# Copyright 2018-2021 Elyra Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from abc import ABC
+from abc import abstractmethod
+from enum import Enum
+from http import HTTPStatus
+import os
+import re
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
+from urllib.parse import urlsplit
+
+import requests
+
+
+class SupportedKFPAuthProviders(Enum):
+    """
+    List of supported authentication providers that is defined
+    in this module. Each entry in this list must be associated
+    with an implementation of AbstractKFPAuthenticator. Each
+    entry's value must also be specified as enum value for the
+    "auth_type" property in elyra/kfp/metadata/schemas/kfp.json
+    to enable users to select the provider from the list.
+
+    """
+    # KF is not secured
+    # (See UnsecuredKFPAuthenticator)
+    NO_AUTHENTICATION = 'No authentication'
+    # KF is secured using static id/password
+    # (See StaticPasswordKFPAuthenticator implementation)
+    STATIC_PASSWORD = 'Static password'
+    # KF is secured using KF_PIPELINES_SA_TOKEN_PATH
+    # (See SATokenPathKFPAuthenticator implementation)
+    KF_PIPELINES_SA_TOKEN_PATH = 'KF_PIPELINES_SA_TOKEN_PATH'
+    # Supports multiple authentication mechanisms
+    # (See AutoKFPAuthenticator implementation)
+    AUTO = 'Auto'
```

Comment:
I strongly think this should NOT be called `AUTO`, as this implies we are doing some kind of authentication-type detection. 

If you really want to keep the old authentication code in Elyra, the only sensible name would be `DEX_LEGACY`.

__NOTE:__ the code for `AUTO` right now is actually just a less sophisticated version of the `DEX_STATIC_PASSWORDS` auth (It won't work if multiple Dex providers are enabled).

Reply:
I changed the display label to `DEX (generic)` and added information to the documentation when to choose this type.